### PR TITLE
[APB-4427] Personal Details Validation journey for individuals with no NINO

### PR DESCRIPTION
**Pull request by stevenmallen**

### CODE REVIEW

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/testing/TestEndpointsController.scala [hunk 1]
```diff
@@ -16,7 +16,9 @@
 
 package uk.gov.hmrc.agentinvitationsfrontend.controllers.testing
 
-import javax.inject.Inject
+import java.net.URL
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/controllers/testing/TestEndpointsController.scala]: are we using this import ?

File: app/uk/gov/hmrc/agentinvitationsfrontend/config/ExternalUrls.scala [hunk 2]
```diff
@@ -37,7 +37,8 @@ class ExternalUrls @Inject()(
   @Named("privacy-policy.external-url") val privacypolicyUrl: String,
   @Named("agent-invitations-frontend.external-url") val agentInvitationsExternalUrl: String,
   @Named("vat-online-service-helpline.external-url") val vatOnlineServiceHelplineUrl: String,
-  @Named("sa-online-service-helpline.external-url") val saOnlineServiceHelplineUrl: String
+  @Named("sa-online-service-helpline.external-url") val saOnlineServiceHelplineUrl: String,
+  @Named("personal-details-validation-frontend.external-url") val personalDetailsValidationFrontendExternalUrl: String
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/config/ExternalUrls.scala]: can we use a shorter name for the val like `pdvFrontendUrl` ?

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala [hunk 3]
```diff
@@ -60,4 +65,23 @@ class IdentityVerificationConnector @Inject()(
...
+    implicit hc: HeaderCarrier,
+    ec: ExecutionContext): Future[Int] =
+    monitor("ConsumedAPI-Client-updateEntry-PUT") {
+      http
+        .PUT[NinoClStoreEntry, HttpResponse](updateEntryUrl(credId).toString, entry)
+        .map(_.status)
+        .recover {
+          case e: Upstream5xxResponse => {
+            Logger.error(s"identity-verification did not update entry ${e.getMessage}")
+            Status.INTERNAL_SERVER_ERROR
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: in general its not a good idea to deal with `Results` in connectors - 
in controllers we can return `Result` based on the connector/service response

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala [hunk 4]
```diff
@@ -22,23 +22,28 @@ import com.codahale.metrics.MetricRegistry
...
 
 import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
 class IdentityVerificationConnector @Inject()(
-  @Named("identity-verification-frontend-baseUrl") baseUrl: URL,
-  http: HttpGet,
+  @Named("identity-verification-frontend-baseUrl") ivFrontendBaseUrl: URL,
+  @Named("identity-verification-baseUrl") ivBaseUrl: URL,
+  http: HttpGet with HttpPut,
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: why do we use HttpClient from bootstrap here, that gives get, put, post..etc

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class PersonalDetailsValidationConnector @Inject()(
+  @Named("personal-details-validation-baseUrl") baseUrl: URL,
+  metrics: Metrics,
+  http: HttpGet,
+  env: Environment,
+  config: Configuration)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: minor: but all the above 3 params could go into an implicit list

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala [hunk 6]
```diff
@@ -60,4 +65,23 @@ class IdentityVerificationConnector @Inject()(
...
     }
+
+  /** Call identity-verification to update NINO store
+    * @return The HTTP status code of the PUT response
+    *
+    * */
+  def updateEntry(entry: NinoClStoreEntry, credId: String)(
+    implicit hc: HeaderCarrier,
+    ec: ExecutionContext): Future[Int] =
+    monitor("ConsumedAPI-Client-updateEntry-PUT") {
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: this doesn't give us much context, can we use ninoUpdateEntry or IVUpdateEntry something like that ?  if not no issues

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+          case e: NotFoundException => {
+            Logger.warn(s"personal details validation did not recognise the validation is $validationId $e")
+            Left(PdvValidationNotFound)
+          }
+        }
+    }
+}
+
+sealed trait PdvError {
+  def Result[A](validationId: String)(
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: same as above for returning Result from connectors..

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    implicit request: Request[A],
+    messages: Messages,
+    configuration: Configuration,
+    externalUrls: ExternalUrls): Result = this match {
+    case PdvValidationNotFound =>
+      InternalServerError(s"failed to get PDV result: data for validationId $validationId not found")
+    case PdvValidationNoNino  => InternalServerError(s"failed to get PDV result: No NINO in response for $validationId")
+    case PdvValidationFailure => Forbidden(cannot_confirm_identity())
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: i can see the drawback already:  dealing with views in controllers

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala [hunk 9]
```diff
@@ -186,7 +187,7 @@ class ClientInvitationJourneyController @Inject()(
   }
 
   def incorrectlyAuthorisedAsAgent: Action[AnyContent] = Action.async { implicit request =>
-    withAuthorisedAsAgent { _ =>
+    authActions.withAuthorisedAsAgent { _ =>
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala]: do we really need this prefix `authActions.` ?

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala [hunk 10]
```diff
@@ -215,6 +216,42 @@ class ClientInvitationJourneyController @Inject()(
...
+              .getPdvResult(validId)
+              .flatMap {
+                case Right(nino) =>
+                  identityVerificationConnector
+                    .updateEntry(NinoClStoreEntry(providerId, nino, None, None, None), providerId)
+                    .map {
+                      case 200 | 201 => Redirect(targetUrl)
+                      case status =>
+                        Logger.error(s"identity-verification upsert /nino/credId returned: $status")
+                        InternalServerError("identity-verification upsert NINO failed")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala]: if we get a 400 back do we still throw ISE here, if we dont get a 4xx then its fine,

File: it/uk/gov/hmrc/agentinvitationsfrontend/support/BaseISpec.scala [hunk 11]
```diff
@@ -1,6 +1,6 @@
 package uk.gov.hmrc.agentinvitationsfrontend.support
 
-import java.net.URLEncoder
+import java.net.{URL, URLEncoder}
```
Comment (sureshhmrc) [inline it/uk/gov/hmrc/agentinvitationsfrontend/support/BaseISpec.scala]: i think unused import, but we can ignore not a big deal

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/testing/TestEndpointsController.scala (hunk 1 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/controllers/testing/TestEndpointsController.scala]: Nope - will remove

File: app/uk/gov/hmrc/agentinvitationsfrontend/config/ExternalUrls.scala (hunk 2 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/config/ExternalUrls.scala]: I was thinking the same thing!

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala (hunk 3 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: Actually, we are just returning the HTTP response status (Int), so I think this is OK?

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala (hunk 4 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: I think this was Geoff's choice - not sure what difference it makes?

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala (hunk 5 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: what is the advantage of doing this?

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala (hunk 6 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: Yep, will change

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala (hunk 7 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: I'll move this back to the controller; this is where I put it initially.

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala (hunk 8 above)
Comment (JTY009) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: Define drawback? I am curious to know what is the problem? apart from breaking consistence to some extent

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala (hunk 9 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala]: Nope, can remove

File: app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala (hunk 10 above)
Comment (stevenmallen) [reply app/uk/gov/hmrc/agentinvitationsfrontend/controllers/ClientInvitationJourneyController.scala]: Anything other than 200 or 201 is a technical error.

File: it/uk/gov/hmrc/agentinvitationsfrontend/support/BaseISpec.scala [hunk 12]
```diff
@@ -82,6 +83,12 @@ abstract class BaseISpec
         "microservice.services.sso.port"                                      -> wireMockPort,
         "microservice.services.identity-verification-frontend.host"           -> wireMockHost,
         "microservice.services.identity-verification-frontend.port"           -> wireMockPort,
+        "microservice.services.identity-verification.host"                    -> wireMockHost,
+        "microservice.services.identity-verification.port"                    -> wireMockPort,
+        "microservice.services.personal-details-validation.host"              -> wireMockHost,
+        "microservice.services.personal-details-validation.port"              -> wireMockPort,
+        "microservice.services.personal-details-validation-frontend.host"     -> wireMockHost,
```
Comment (JTY009) [inline it/uk/gov/hmrc/agentinvitationsfrontend/support/BaseISpec.scala]: I don't think we have PDVF host and port in application.conf anymore

Comment (stevenmallen) [reply it/uk/gov/hmrc/agentinvitationsfrontend/support/BaseISpec.scala]: Good spot!

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala (hunk 4 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: if we use httpClient from bootstrap, we get all get/post/put/delete for free so you don't have to add the `with Http*` each time you need a *
Also, I remember Auding hooks comes by default with DefaultHttpClient , but anyway not a big deal if you don't have time for making those changes

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala (hunk 3 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/IdentityVerificationConnector.scala]: ah yes, returning int/status is fine

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala (hunk 5 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: i always prefer to declare off-business logic deps in an implicit list like metrics, Env, Config, ExecutionContext... etc as these are not direct dependencies that our business logic depends on... 
the main advantage comes when we have unit tests around these...(let's discuss later), but for now these are no harm

File: app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala (hunk 8 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentinvitationsfrontend/connectors/PersonalDetailsValidationConnector.scala]: connectors are only responsible for making HTTP calls and retuning status/response, they do not deal with play.api.Results or HTML...not even auditing or monitoring stuff... unfortunately our connectors do a lot more than making HTTP calls :) 
What if some other service wants to use this connector's method and want to tackle the HTTP status/response in a different way?
It's controllers responsibility to return a Result/view to the caller

**Pull request merged**